File: source/python/144. Binary Tree Preorder Traversal.py
from queue import LifoQueue
from typing import Self


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val: int = 0, left: Self | None = None, right: Self | None = None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    # def preorderTraversal(self, root: TreeNode | None) -> list[int]:
    #     if root is None:
    #         return []
    #     return [root.val] + self.preorderTraversal(root.left) + self.preorderTraversal(root.right)

    def preorderTraversal(self, root: TreeNode | None) -> list[int]:
        if root is None:
            return []

        stack: LifoQueue[TreeNode] = LifoQueue(maxsize=100)
        stack.put(root)

        result: list[int] = []

        while not stack.empty():
            node = stack.get()
            result.append(node.val)

            if node.right is not None:
                stack.put(node.right)
            if node.left is not None:
                stack.put(node.left)

        return result


def test():
    s = Solution()
    assert s.preorderTraversal(
        TreeNode(1, None, TreeNode(2, TreeNode(3)))) == [1, 2, 3]
    assert s.preorderTraversal(None) == []
    assert s.preorderTraversal(TreeNode(1)) == [1]

    print("All tests passed!")


if __name__ == "__main__":
    test()

File: source/cpp/144. Binary Tree Preorder Traversal.cpp
#include <stack>
#include <vector>


// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    std::vector<int> preorderTraversal(const TreeNode* root) const {
        if (root == nullptr) return {};

        std::vector<int> result;
        std::stack<const TreeNode*> stack;
        stack.push(root);

        while (!stack.empty()) {
            auto node = stack.top();
            stack.pop();
            result.push_back(node->val);

            if (node->right != nullptr) stack.push(node->right);
            if (node->left != nullptr) stack.push(node->left);
        }

        return result;
    }

    std::vector<int> ans;
    void preorder(const TreeNode* root) {
        if(root==nullptr) return;

        ans.push_back(root->val);
        preorder(root->left);
        preorder(root->right);
    }
    std::vector<int> preorderTraversal(const TreeNode* root) {
        //root left right
        preorder(root);
        return ans;
    }
};
